### PR TITLE
Support IPROTO_PUSH messages

### LINKED ISSUE #67: Support IPROTO_PUSH
box.session.push() in 1.10 impements a new iproto command, IPROTO_PUSH. Please add support for it.

### PR DESCRIPTION
**Pull request by oleg-jukovec**
This patch adds support for receiving messages sent using box.session.push()
via an iterator in the manner of asynchronous case of a Lua implementation[1].

Now the calls Future.Get() and Future.GetTyped() ignore push messages, and
do not report an error.

1. https://www.tarantool.io/ru/doc/latest/reference/reference_lua/box_session/push/

Closes #67

I'm not particularly familiar with Go. So I most likely made mistakes in some basic and idiomatic things

### CODE REVIEW

File: config.lua
Comment (ligurio) [inline config.lua:60]: Missed whitespaces around `=`

See Lua style guide https://www.tarantool.io/en/doc/latest/dev_guide/lua_style_guide/

**Review by ligurio — CHANGES_REQUESTED** [review]

File: tarantool_test.go
Comment (oleg-jukovec) [reply tarantool_test.go:696]: I added ExampleFuture_GetIterator() into example_test.go

File: tarantool_test.go
Comment (oleg-jukovec) [reply tarantool_test.go]: Fixed, thanks!

File: example_test.go
Comment (DifferentialOrange) [inline example_test.go:96]: One of purposes of the push is to be able to send interactive signals while processing something long.

So instead of a function that tries to push everything rather quickly and return a result I've tried to use something slower:
```lua
local fiber = require('fiber')

local function push_func(x)
    for i = 1, x do
        box.session.push(i)
        fiber.sleep(0.3) -- 0.3 seconds
    end
    return x
end
```

I failed to run the example:
```bash
go clean -testcache && go test -run 'ExampleFuture_GetIterator' -timeout 100s
--- FAIL: ExampleFuture_GetIterator (0.50s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x55e70b]

goroutine 1 [running]:
testing.(*InternalExample).processRunResult(0xc00013dae8, {0x0, 0x0}, 0xc0001cc018?, 0x0, {0x586840, 0x6dc270})
	/usr/local/go/src/testing/example.go:91 +0x4e5
testing.runExample.func2()
	/usr/local/go/src/testing/run_example.go:59 +0x11c
panic({0x586840, 0x6dc270})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/tarantool/go-tarantool_test.ExampleFuture_GetIterator()
	/home/georgymoiseev/Development/github/tarantool/go-tarantool/example_test.go:103 +0x1cb
testing.runExample({{0x5b653f, 0x19}, 0x5beef0, {0x5ba2e4, 0x24}, 0x0})
	/usr/local/go/src/testing/run_example.go:63 +0x28d
testing.runExamples(0xc00013dcd8, {0x6e10e0?, 0x4, 0x5?})
	/usr/local/go/src/testing/example.go:44 +0x186
testing.(*M).Run(0xc00011c460)
	/usr/local/go/src/testing/testing.go:1721 +0x689
github.com/tarantool/go-tarantool_test.runTestMain(0x1?)
	/home/georgymoiseev/Development/github/tarantool/go-tarantool/tarantool_test.go:1129 +0x1b5
github.com/tarantool/go-tarantool_test.TestMain(0x4413f1?)
	/home/georgymoiseev/Development/github/tarantool/go-tarantool/tarantool_test.go:1133 +0x19
main.main()
	_testmain.go:85 +0x1d3
exit status 2
FAIL	github.com/tarantool/go-tarantool	0.612s
```

I don't know if its some test timeout, but the only one documented `go test` timeout I've found is a `-timeout` flag which is set to be > 3s, so I suppose it's not the test framework fault.

Running iterator fetching cycle with 3s timeout and failing to process a 1.2s function is rather disappointing. It should either be fixed in the inside so it would be supported or described on the outside how one should use timeouts with long requests because for not it is not obvious (or not works).

**Review by DifferentialOrange — COMMENTED** [review]
I've tried to run the example with "slow" function and I failed to do it successfully, even though timeout is ok. See comment below for additional details.

File: request.go
Comment (oleg-jukovec) [reply request.go:474]: I think indefinite article is correct here because the response is not known yet in this context/comment.

File: example_test.go
Comment (oleg-jukovec) [reply example_test.go:97]: Thanks, I fixed the crash and a reason of it.

Indeed, this is an implementation limitation. It makes no sense to wait for a response longer than conn.opts.Timeout. It has [performance purposes](https://github.com/tarantool/go-tarantool/issues/48#issuecomment-334364328), so I don't change it. I have updated the comment for TimeoutResponseIterator.WithTimeout() with a note.

You can to test longer timeouts in a such way:
```
--- a/config.lua
+++ b/config.lua
@@ -56,10 +56,12 @@ local function simple_incr(a)
 end
 rawset(_G, 'simple_incr', simple_incr)
 
+fiber = require("fiber")
 local function push_func(cnt)
     for i = 1, cnt do
         box.session.push(i)
     end
+       fiber.sleep(1)
     return cnt
 end
 rawset(_G, 'push_func', push_func)
diff --git a/tarantool_test.go b/tarantool_test.go
index a9c0207..4243592 100644
--- a/tarantool_test.go
+++ b/tarantool_test.go
@@ -90,7 +90,7 @@ var spaceName = "test"
 var indexNo = uint32(0)
 var indexName = "primary"
 var opts = Opts{
-       Timeout: 500 * time.Millisecond,
+       Timeout: 5 * time.Second,
        User:    "test",
        Pass:    "test",
        //Concurrency: 32,

```

Comment (oleg-jukovec) [reply example_test.go:97]: Fixed, thanks!

File: example_test.go
Comment (oleg-jukovec) [reply example_test.go]: Updated, thank you!

File: request.go
Comment (oleg-jukovec) [reply request.go]: It is copied from Future.Get() description:

"Response will contain deserialized result in Data field."

I think it's correct, but I have updated the comment %s/will contain result/will contain deserialized result/g

File: request.go
Comment (oleg-jukovec) [reply request.go]: Fixed, thanks!

File: response_it.go
Comment (oleg-jukovec) [reply response_it.go]: I don't see a critical difference, but I have updated it.

File: response_it.go
Comment (oleg-jukovec) [reply response_it.go]: Fixed, thanks!

File: config.lua
Comment (oleg-jukovec) [reply config.lua]: Replaced by "cnt"

File: config.lua
Comment (ligurio) [inline config.lua:56]: Please rebase to master branch and add changelog entry.

**Review by ligurio — CHANGES_REQUESTED** [review]

File: config.lua
Comment (oleg-jukovec) [reply config.lua:56]: Thanks, done.

File: example_test.go
Comment (oleg-jukovec) [reply example_test.go:99]: Why not do it right in the Future.GetIterator() comment?The generated doc will look nice to my taste:

link

But of course I can try something else if you don't agree. Do you want a more global mention of box.session.push() ?

Comment (ligurio) [reply example_test.go:99]: It's ok for my taste too, let's keep it as is.

**Review by ligurio — DISMISSED** [review]

File: example_test.go
Comment (oleg-jukovec) [reply example_test.go]: Updated.

File: example_test.go
Comment (DifferentialOrange) [reply example_test.go:110]: I think it would be nice to go a bit further and log a warning when user calls `fut.GetIterator().WithTimeout(timeout)` with a timeout greater than request timeout. Request timeout is stored in `conn.opts.Timeout` and there are some log examples already: https://github.com/tarantool/go-tarantool/blob/a535b8e0224a63869dc911e8c39e72844929b322/connection.go#L65-L81

**Review by DifferentialOrange — COMMENTED** [review]
Thank you for your responses and fixes.

I have left some comments. There shouldn't be anything critical this time. I think it will be the last round and I'll approve the PR after we deal with these comments.

File: tarantool_test.go
Comment (oleg-jukovec) [reply tarantool_test.go:674]: This is a copy-paste of the used pattern for connection in almost all tests. We need to refactor it and add a helper function such that:
```go
func connect(t testing TB, server string, opts Opts) {
    t.Helper()

    conn, err := Connection(server, opts)
    if err != nil {
        t.Fatal(...)
    }
    if conn != nil {
        t.Fatal(...)
    }
}
```
This needs to be done for all tests. But it is out of scope ot the issue (we need another one). So I didn't try to fix the connection pattern in only tests for the issue.

File: response_it.go
Comment (funny-falcon) [inline response_it.go:54]: I don't get:
- if iterator doesn't get timeout parameter
- then it doesn't wait anything
- therefore usually it doesn't get anything, because nextResponse sees empty `fut.pushes` and empty `resp`.
LOL
What i'm missing?

File: response_it.go
Comment (funny-falcon) [inline response_it.go:71]: strange decision to set wg.Done() so early.
Why not simply `defer wg.Done()` ?

File: request.go
Comment (funny-falcon) [inline request.go:17]: Instead of (ab)using sync.Cond, you'd better
```
    // this doesn't consume more memory, than make(chan struct{})
    fut.ready := make(chan struct{}, 1000000000) 
```
And then notify about PUSH using `fut.ready <- struct{}{}`
This way you will need no separate goroutine to merry push notifications with timeouts and final response.
Whole iterator handling will be much simpler.

**Review by funny-falcon — CHANGES_REQUESTED** [review]
https://github.com/tarantool/go-tarantool/pull/156/files#r855318180

File: response_it.go
Comment (oleg-jukovec) [reply response_it.go:54]: Exactly, see Lua implementation:
https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_session/push/
Lua's iterator just leave a loop if nothing to do (returns it.Next() == false in our case).
```
-- There is no pause because conn:call does not wait for
-- server_function to finish. The first time that we go through
-- the pairs() loop, we see the messages table is empty. Like this:
-- tarantool> messages
```

File: tarantool_test.go
Comment (oleg-jukovec) [reply tarantool_test.go]: Fixed, thanks!

File: tarantool_test.go
Comment (oleg-jukovec) [reply tarantool_test.go:692]: Fixed, thanks!

File: request.go
Comment (oleg-jukovec) [reply request.go:17]: Yes, it will be easy. But the code is similar to the Lua implementation will be incorrect:
https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_session/push/
```Lua
for i, message in future:pairs() do
    table.insert(messages, message) table.insert(keys, i) end
messages
future:wait_result(1000)
for i, message in future:pairs() do
    table.insert(messages, message) table.insert(keys, i) end
messages
```

File: response_it.go
Comment (funny-falcon) [reply response_it.go:53]: Go is not Lua. It doesn't need to match Lua's code exactly.
In Go one gets notifications through channels. Channels are composable.
"Polling" on `it.Next()` would be definitely anti-pattern in Go.

Well... I could not came with clear api now.
It doesn't look like iterator could be efficiently implemented in Go-friendly way if there is no separate method "conn.CallStream".

File: tarantool_test.go
Comment (DifferentialOrange) [reply tarantool_test.go:674]: >We need to refactor it

Yes, this is what https://github.com/tarantool/go-tarantool/issues/150 is about. I think it wouldn't be bad to be inconsistent with existing test here and do the right thing this time.

File: example_test.go
Comment (oleg-jukovec) [reply example_test.go:110]: > Shouldn't we reset future's timeout (and so move it to the tail of requests list due to the way timeouts implemented)?

It seems to me that this is logical. I have updated the peekFuture function:

```patch
--- a/connection.go
+++ b/connection.go
@@ -754,15 +773,34 @@ func (conn *Connection) putFuture(fut *Future, body func(*msgpack.Encoder) error
        }
 }
 
+func (conn *Connection) peekFuture(reqid uint32) (fut *Future) {
+       shard := &conn.shard[reqid&(conn.opts.Concurrency-1)]
+       pos := (reqid / conn.opts.Concurrency) & (requestsMap - 1)
+       shard.rmut.Lock()
+       defer shard.rmut.Unlock()
+
+       if conn.opts.Timeout > 0 {
+               fut = conn.getFutureImp(reqid, true)
+               pair := &shard.requests[pos]
+               *pair.last = fut
+               pair.last = &fut.next
+               fut.timeout = time.Now().Sub(epoch) + conn.opts.Timeout
+       } else {
+               fut = conn.getFutureImp(reqid, false)
+       }
+
+       return fut
+}
+
```

File: tarantool_test.go
Comment (oleg-jukovec) [reply tarantool_test.go]: It's okay here. The code below should process error cases correctly.

File: tarantool_test.go
Comment (oleg-jukovec) [reply tarantool_test.go]: Updated.

File: tarantool_test.go
Comment (oleg-jukovec) [reply tarantool_test.go]: It's okey here. The code below should process error cases correctly.

File: request.go
Comment (oleg-jukovec) [reply request.go]: Thank you, it's much easier.

File: tarantool_test.go
Comment (oleg-jukovec) [reply tarantool_test.go]: And what are we going to test? That count of responses (push messages + a response) >= 0 && <= cnt + 1? I added such test with iteration without timeout && fiber.sleep().
```patch
--- a/tarantool_test.go
+++ b/tarantool_test.go
@@ -632,12 +632,36 @@ func TestClientSessionPush(t *testing.T) {
        conn := connect(t, server, opts)
        defer conn.Close()
 
+       var it ResponseIterator
        const pushMax = 3
+       // It will iterate immediately.
+       fut0 := conn.Call17Async("push_func", []interface{}{pushMax})
+       respCnt := 0
+       for it = fut0.GetIterator(); it.Next(); {
+               err := it.Err()
+               resp := it.Value()
+               if err != nil {
+                       t.Errorf("Unexpected error after it.Next() == true: %q", err.Error())
+                       break
+               }
+               if resp == nil {
+                       t.Errorf("Response is empty after it.Next() == true")
+               }
+               respCnt += 1
+       }
+       if err := it.Err(); err != nil {
+               t.Errorf("An unexpected iteration error: %s", err.Error())
+       }
+       if respCnt < 0 || respCnt > pushMax + 1 {
+               t.Errorf("Unexpected respCnt = %d, expected 0 <= respCnt <= %d", respCnt, pushMax + 1)
+       }
```

Our tests depends on Opts.Timeout value. They can sometimes fail (CI/CD container was paused by a time > Opts.Timeout) and it's ok. In fact in our tests we can 'guarantee' only that count of responses (push messages + a response) will be >= 0, nothing more. If we add more timeouts to our tests, it will only be more difficult to 'guarantee' their work when we expect it.

Comment (DifferentialOrange) [reply tarantool_test.go]: > Is count of responses (push messages + a response) >= 0 && <= cnt + 1?

Isn't `it.Next()` waiting until next push value?

Well, if timeouts are adequate the possibility of tests failing because of some external disturbance is really low (I haven't seen any issues in this repo since we introduced CI/CD). You may increase them here if you're wish.

Green tests are good, but it is even better to have tests that replicate some real case. And I think that "long" request with pushes is closer to reality than consecutive pushing in cycle without any other work. It is highly likely that this "non-ready" request in fact is a "ready" request because Lua procedure is really fast and test may not cover the case we're expecting.

File: example_test.go
Comment (DifferentialOrange) [reply example_test.go:110]: Disregarding the approach we will implement here, we should document it in detail because it is definitely the place that may leave users confused.

File: tarantool_test.go
Comment (oleg-jukovec) [reply tarantool_test.go]: No, it doesn't wait if no timeout. See an example from https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_session/push/ for an asynchronous case.

It still won't give guarantees. Opts.Timeout = 1s, 2s or 10m just increase the chance of a successful result.

And yet, it seems to me timeouts and sleeps is very bad idea for unit tests. I suggest testing the iterator in a different way:

Move Future.appendPushes/Future.setResponse/Future.setError into the public API. And write unit-tests for Future + Future.GetResponseIterator() without Connection, Tarantool etc.

Comment (oleg-jukovec) [reply tarantool_test.go]: > Green tests are good, but it is even better to have tests that replicate some real case

Please see future_test.go after the update. Yes, we still don't test connection timeouts (we don't do that anywhere). But the iterator itself is tested for this case. That may be enough.

Much more refactoring will be needed to test connection timeouts (and we need to think about how to do it). It seems to me that it is better to do this in a separate issue.

File: example_test.go
Comment (oleg-jukovec) [reply example_test.go:110]: > Disregarding the approach we will implement here, we should document it in detail because it is definitely the place that may leave users confused.

I have added notes to Opts.Timeout and TimeoutResponseIterator.WithTimeout(). I think, the user will quickly notice that something is going wrong and look at the documentation.

Logging is also good. But it will be ugly in the current code. Future should know about Connection's params (Logger, Timeout)... It will be better to comment =)

File: connection.go
Comment (DifferentialOrange) [inline connection.go:161]:
```suggestion
	// Timeout for response to a particular request. The timeout is reset when
```

File: connection.go
Comment (DifferentialOrange) [inline connection.go:162]: "If Timeout is zero request" -> "If Timeout is zero"?

File: example_test.go
Comment (DifferentialOrange) [reply example_test.go:110]: Ok, I don't have any more questions

File: future_test.go
Comment (DifferentialOrange) [inline future_test.go:47]:
```suggestion
		t.Errorf("An unexpected next value.")
```

File: future_test.go
Comment (DifferentialOrange) [inline future_test.go:61]:
```suggestion
			t.Errorf("An unexpected next value.")
```

File: future_test.go
Comment (DifferentialOrange) [inline future_test.go:27]:
```suggestion
		t.Errorf("Unexpected response %v, expected %v", it.Value(), resp)
```

File: future_test.go
Comment (DifferentialOrange) [inline future_test.go:136]:
```suggestion
			t.Errorf("An unexpected next value.")
```

File: future_test.go
Comment (DifferentialOrange) [inline future_test.go:156]:
```suggestion
		t.Errorf("An error expeceted.")
```

File: future_test.go
Comment (DifferentialOrange) [inline future_test.go:170]:
```suggestion
		if i == len(responses) - 1 {
```

File: future_test.go
Comment (DifferentialOrange) [inline future_test.go:225]:
```suggestion
				t.Errorf("an unexpected error %q, expected %q", err.Error(), errMsg)
```

File: connection.go
Comment (oleg-jukovec) [reply connection.go]: Thank you.
```patch
-       // Timeout for response to a particular request. The timout is reset when
-       // push messages are received. If Timeout is zero request, any request can
-       // be blocked infinitely.
+       // Timeout for response to a particular request. The timeout is reset when
+       // push messages are received. If Timeout is zero, any request can be
+       // blocked infinitely.
```

File: connection.go
Comment (oleg-jukovec) [reply connection.go]: Fixed, thanks.

File: future_test.go
Comment (oleg-jukovec) [reply future_test.go:170]: It's ok. ```go fmt``` did it =)

File: future_test.go
Comment (DifferentialOrange) [inline future_test.go:131]: Is this situation possible in real life?

File: future_test.go
Comment (DifferentialOrange) [inline future_test.go:150]: Is this situation possible in real life? [2]

File: tarantool_test.go
Comment (DifferentialOrange) [reply tarantool_test.go]: Future unit tests are detailed and should cover all situations including one I've wanted do. Thank you, I think then this issue is resolved.

**Review by DifferentialOrange — APPROVED** [review]
Should be LGTM after resolving remaining comments.

Please don't forget to rebase onto master if needed (for example, when https://github.com/tarantool/go-tarantool/pull/160 will be merged) and solve all possible conflicts and warnings.

File: future_test.go
Comment (oleg-jukovec) [reply future_test.go:131]: Not now. But this test determines the behavior of Future for the case. So I think it's ok.

File: future_test.go
Comment (oleg-jukovec) [reply future_test.go:150]: Not now. But this test determines the behavior of Future for the case. So I think it's ok. [2]

**Review by DifferentialOrange — APPROVED** [review]
AFAIR, there shouldn't be any unsolved questions, isn't it?

**Comment by oleg-jukovec** [comment]
> AFAIR, there shouldn't be any unsolved questions, isn't it?

For me it seems not, I only sometimes do rebase.

**Pull request merged**
